src/Tools/dotnet-monitor/CollectionRules/CollectionRulePipeline.cs

```diff
@@ -210,16 +216,58 @@ protected override async Task OnRun(CancellationToken token)
                         // Complete the pipeline since the action list is only executed once
                         // for collection rules with startup triggers.
                         completePipeline = true;
+
+                        StateHolder.StartupTriggerCompleted();
                     }
                 }
             }
             catch (OperationCanceledException) when (durationCancellationSource.IsCancellationRequested)
             {
                 // This exception is caused by the pipeline duration expiring.
                 // Handle it to allow pipeline to be in completed state.
+                StateHolder.RuleDurationReached();
+            }
+            catch (Exception)
+            {
+                StateHolder.RuleFailure(); // Untested...not sure if this is correct behavior
+                throw;
+            }
+        }
+
+        public static void DequeueOldTimestamps(Queue<DateTime> timestamps, TimeSpan? actionCountWindowDuration, DateTime currentTimestamp)
+        {
+            // If rule has an action count window, remove all execution timestamps that fall outside the window.
+            if (actionCountWindowDuration.HasValue)
+            {
+                DateTime windowStartTimestamp = currentTimestamp - actionCountWindowDuration.Value;
+                while (timestamps.Count > 0)
+                {
+                    DateTime executionTimestamp = timestamps.Peek();
+                    if (executionTimestamp < windowStartTimestamp)
+                    {
+                        timestamps.Dequeue();
+                    }
+                    else
+                    {
+                        // Stop clearing out previous executions
+                        break;
+                    }
+                }
             }
         }
 
+        internal bool CheckForThrottling(int actionCountLimit, int currExecutions, TimeSpan? actionCountSWD)
+        {
+            if (!actionCountSWD.HasValue || actionCountLimit > currExecutions)
+            {
+                StateHolder.EndThrottled();
```

Comment:
My concern here is that we call this from several places, which means each caller would be responsible for having a state-change directly after it, which just feels like it's decoupling two things that are inherently coupled. There was a prior iteration of this that used cancellation tokens to get instant updates instead of manually calling this method to check if things are up-to-date, but I know Justin mentioned that would make testing using simulated time more challenging.